pkg/services/grpcserver/service.go

```diff
@@ -0,0 +1,98 @@
+package grpcserver
+
+import (
+	"context"
+	"fmt"
+	"net"
+
+	"github.com/grafana/grafana-plugin-sdk-go/backend"
+	"github.com/grafana/grafana/pkg/infra/log"
+	"github.com/grafana/grafana/pkg/registry"
+	"github.com/grafana/grafana/pkg/services/apikey"
+	"github.com/grafana/grafana/pkg/services/featuremgmt"
+	"github.com/grafana/grafana/pkg/services/sqlstore"
+	"github.com/grafana/grafana/pkg/setting"
+	grpcAuth "github.com/grpc-ecosystem/go-grpc-middleware/auth"
+
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials"
+	"google.golang.org/grpc/reflection"
+)
+
+type Provider interface {
+	registry.BackgroundService
+	GetServer() *grpc.Server
+}
+
+type GPRCServerService struct {
+	cfg    *setting.Cfg
+	logger log.Logger
+	server *grpc.Server
+}
+
+func ProvideService(cfg *setting.Cfg, apiKey apikey.Service, sqlStore sqlstore.Store) (Provider, error) {
+	s := &GPRCServerService{
+		cfg:    cfg,
+		logger: log.New("grpc-server"),
+	}
+
+	var opts []grpc.ServerOption
+
+	// Default auth is admin token check, but this can be overridden by
+	// services which implement ServiceAuthFuncOverride interface.
+	// See https://github.com/grpc-ecosystem/go-grpc-middleware/blob/master/auth/auth.go#L30.
+	authenticator := NewAuthenticator(apiKey, sqlStore)
+	opts = append(opts, []grpc.ServerOption{
+		grpc.StreamInterceptor(grpcAuth.StreamServerInterceptor(authenticator.Authenticate)),
+		grpc.UnaryInterceptor(grpcAuth.UnaryServerInterceptor(authenticator.Authenticate)),
+	}...)
+
+	if s.cfg.GRPCServerTLSConfig != nil {
+		opts = append(opts, grpc.Creds(credentials.NewTLS(cfg.GRPCServerTLSConfig)))
+	}
+
+	grpcServer := grpc.NewServer(opts...)
+	reflection.Register(grpcServer)
```

Comment:
what's the purpose of this line - is just `server.run()`?

Reply:
i vaguely remember this, but i'm not sure if i added it. it's for [GRPC server reflection](https://github.com/grpc/grpc/blob/master/doc/server-reflection.md)